src/App.tsx

```diff
@@ -15,48 +16,122 @@ export const goodsFromServer = [
+}
+
+type ReorderOptions = {
+  sortType: SortType,
+  isReversed: boolean,
+};
+
+// Use this function in the render method to prepare goods
+export function getReorderedGoods(
+  goods: string[],
+  { sortType, isReversed }: ReorderOptions,
+) {
+  // To avoid the original array mutation
+  const visibleGoods = [...goods];
+
+  // Sort and reverse goods if needed
+  visibleGoods.sort((el1, el2) => {
+    switch (sortType) {
+      case SortType.NONE:
+        return 0;
+
+      case SortType.ALPHABET:
+        return el1.localeCompare(el2);
+
+      case SortType.LENGTH:
+        return el1.length - el2.length;
+
+      default:
+        return 0;
+    }
+  });
+
+  if (isReversed) {
+    visibleGoods.reverse();
+  }
+
+  return visibleGoods;
+}
+
 export const App: React.FC = () => {
+  const [isReversed, setIsReversed] = useState(false);
+  const [sortType, setSortType] = useState(SortType.NONE);
+
+  const reversed = () => {
+    setIsReversed(prevState => !prevState);
+  };
+
+  const reset = () => {
+    setIsReversed(false);
+    setSortType(SortType.NONE);
+  };
+
+  const sortByLength = () => {
+    setSortType(SortType.LENGTH);
+  };
+
+  const sortAlphabetically = () => {
+    setSortType(SortType.ALPHABET);
+  };
+
+  const showResetBtn = sortType === SortType.NONE && isReversed === false;
+  const classBtnInfo = classNames('button', 'is-info',
+    { 'is-light': sortType !== SortType.ALPHABET });
+  const classBtnSuccess = classNames('button', 'is-success',
+    { 'is-light': sortType !== SortType.LENGTH });
+  const classBtnWarning = classNames('button', 'is-warning',
+    { 'is-light': !isReversed });
+
   return (
     <div className="section content">
       <div className="buttons">
         <button
           type="button"
-          className="button is-info is-light"
+          className={classBtnInfo}
+          onClick={sortAlphabetically}
         >
           Sort alphabetically
         </button>
 
         <button
           type="button"
-          className="button is-success is-light"
+          className={classBtnSuccess}
+          onClick={sortByLength}
         >
           Sort by length
         </button>
 
         <button
           type="button"
-          className="button is-warning is-light"
+          className={classBtnWarning}
+          onClick={reversed}
         >
           Reverse
         </button>
 
         <button
           type="button"
           className="button is-danger is-light"
+          onClick={reset}
+          style={showResetBtn ? { display: 'none' } : {}}
         >
           Reset
         </button>
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {getReorderedGoods(goodsFromServer, { sortType, isReversed })
```

Comment:
don't execute the function from template, move it to the variable

Reply:
still not fixed